src/server.rs

```diff
@@ -396,6 +392,9 @@ impl Server {
     /// This method must be called multiple times while `self.is_data_available()` is true
     /// in order to receive all data the server has to offer.
     pub async fn recv(&mut self) -> Result<BytesMut, Error> {
+        // Our server response buffer. We buffer data before we give it to the client.
+        let mut message_buffer = BytesMut::with_capacity(8196);
```

Comment:
This will allocate a new buffer every time we send a query, this will be much slower than it is now.

Reply:
which is more expensive, cloning or allocating?
cloning does both

Reply:
Why not just pass `self.buffer[..]` here instead avoiding both problems? I just know that in network programming, the buffer is never reallocated and is re-used instead, so we should make sure to do that as well.

To answer your question, I am not sure. My gut feeling makes me thing allocating because it'll be allocated on the heap, but maybe cloning does the same thing in this case.

Reply:
the problem here is we're writing into the server's buffer not the client's send buffer.
also self.buffer[..] returns a different type. The problem is that the buffer isn't shared between the client and server so we can't really populate and flush. Also rust's ownership model makes it difficult too. I can imagine a re-architecture that solves this better but for now this change maintains parity if anything